docs/explanation/locks.rst

```diff
@@ -0,0 +1,236 @@
+Linux kernel locks
+==================
+
+.. note::
+
+  This documentation covers some kernel internals specific to the versions used in Ubuntu real-time kernels.
+  Since internal kernel interfaces are unstable, as discussed in the `"The Linux Kernel Driver Interface" document`_, the details provided here may not be fully accurate for newer kernel versions.
+  However, the development of kernel locks hasn't changed significantly over the years.
+
+
+In the `kernel space`_, certain critical code regions across all `subsystems`_ must be protected from `context switching`_ or, in the case of real-time kernels, from `preemption`_.
+On `SMP systems`_ (Symmetric Multiprocessing), the Linux kernel provides various synchronization mechanisms to safeguard shared data structures from concurrent access, preventing `race conditions`_ in the kernel space.
+
+Locks are primarily used to prevent race conditions, which are a concern in both SMP and real-time systems.
+Even in a single-processor real-time system, the preemption mechanism can lead to race conditions similar to those found in an SMP system.
+
+There are essentially three classes of locks in the kernel:
+
+- `Spinning locks`_
+- `Sleeping locks`_ (also known as blocking locks)
+- `CPU local locks`_
+
+Spinlocks
+----------
+
+Spinlocks (found in `include/linux/spinlock.h`_) are locks that keep polling (or "spinning"), attempting to acquire the lock.
+They are small, fast, and simple to use, making them generally applicable throughout the kernel.
+Spinlocks are ideal for situations where the kernel process cannot sleep.
+The ideal type of spinlock to use depends on the specific contexts between which the data needs to be shared.
+
+Within the class of spinlocks, there are various types:
+
+.. _rw-lock:
+
+rwlock_t
```

Comment:
Could we use the same names to avoid confusion?

Applies to a few other instances.

Reply:
Done! Thank you!